[DOCS] Make ILM documentation data stream aware (#58035)
elastic/elasticsearch: docs/reference/ilm/ilm-tutorial.asciidoc

[discrete]
[[ilm-gs-apply-policy]]
=== Create a composable template to create the data stream and apply the lifecycle policy

> To create a data stream we'll create a composable template with a `data_stream` definition
> that will specify the lifecycle policy.

Reviewer: Nit: Avoid "we" and future tense where possible.

```suggestion
To set up a data stream, first create a composable template to specify the lifecycle policy. Because
the template is for a data stream, it must also include a `data_stream` definition.
```

I wonder if we should mention other requirements here, specifically a timestamp mapping.

Reviewer 2: Starting the sentence with "Because" sounds awkward to me :D my suggestion:
```suggestion
To set up a data stream, first create a composable template to specify the lifecycle policy. A composable template for used a data stream must include a `data_stream` definition.
```

Reviewer: I'm good with Lee's suggestion. Will leave mine up for context.

Author: ```A composable template for used a data stream must include a `data_stream` definition.```
This doesn't make sense to me. Is it meant to be ?
```
For a composable template to be used with a data stream it must include a `data_stream` definition
```

For example, you might create a `timeseries_template` that is will bootstrap a data stream
named `timeseries`.

> To enable the ILM to manage the data stream, the template configures one {ilm-init} setting:

Reviewer: Replaces ILM with attribute.
```suggestion
To enable the {ilm-init} to manage the data stream, the template configures one {ilm-init} setting:
```

* `index.lifecycle.name` specifies the name of the lifecycle policy to apply to the data stream.

You can use the {kib} Create template wizard to add the template.
This wizard invokes the put _index_template API to create the <<indices-templates,composable index template>>
with the options you specify.

The underlying request looks like this:

[source,console]
-----------------------
PUT _index_template/timeseries_template
{

>   "index_patterns": ["timeseries"],                 <1>

Reviewer: Nit: I'd align the callout numbers.

  "data_stream": {
    "timestamp_field": "@timestamp"                  <2>
  },
  "template": {
    "settings": {
      "number_of_shards": 1,
      "number_of_replicas": 1,
      "index.lifecycle.name": "timeseries_policy"     <3>
    }
  }
}

> -----------------------

Reviewer: I think a `date` or `date_nanos` mapping for the timestamp field is now required. I'd included that here.

// TEST[continued]

<1> Apply the template when a document is indexed into the `timeseries` target.

> <2> Identifies the timestamp field for the data source, a field that must be present

Reviewer: Nit: Breaks this into two sentences for readability.
```suggestion
<2> Identifies the timestamp field for the data source. This field must be present
```

in all documents indexed into the `timeseries` data stream.

> <3> The name of the ILM policy used to manage the data stream.

Reviewer: ```suggestion
<3> The name of the {ilm-init} policy used to manage the data stream.
```

You can also invoke this API directly to add templates.

[discrete]

> [[ilm-gs-bootstrap]]
> === Bootstrap the data stream

Reviewer: I'd say "create" rather than "bootstrap" here. Data streams kinda solve the original bootstrap problem so it'd be good to not use the same terminology.
```suggestion
[[ilm-gs-create-the-data-stream]]
=== Create the data stream
```

> To get things started, you simply index a document into the name or wildcard pattern defined

Reviewer: Nit: Use direct imperative.
```suggestion
To get things started, index a document into the name or wildcard pattern defined
```

> in the `index_patterns` of the <<indices-templates,composable index template>> that will
> create the data stream.

Reviewer: Adds some text to note that the target cannot already exist (e.g., I can't create the `logs` data stream if a `logs` index already exists).
```suggestion
in the `index_patterns` of the <<indices-templates,composable index template>>. As long
as an existing data stream, index, or index alias does not already use the name, the index
request automatically creates a corresponding data stream with a single backing index.
```

> {es} will automatically create the corresponding data stream and will index the document into
> the first backing index.

Reviewer: Nit: Avoid future tense.
Also notes that this new index is also the write index.
```suggestion
{es} automatically indexes the request's documents into this backing index, which also
acts as the stream's <<data-stream-write-index,write index>>.
```

For example, the following request creates the `timeseries` data stream and the first generation
backing index called `.ds-timeseries-000001`.

[source,console]
-----------------------
POST timeseries/_doc
{
  "message": "logged the request",
  "@timestamp": "1591890611"
}

-----------------------
// TEST[continued]

> When the rollover conditions are met, the `rollover` action:

Reviewer: I think the conditions use an "OR" logic. This could be read as "rollover happens when condition A _and_ B is met" not "when A _or_ B is met"
```suggestion
When a rollover condition in the data stream's lifecycle policy is met, the `rollover` action:
```

Reviewer 2: I don't like @jrodewig's suggestion, the lifecycle policy does not belong to a data stream (that is a confusing notion). I prefer:
```suggestion
When a rollover condition in the lifecycle policy is met, the `rollover` action:
```

Reviewer: Good point. I'll leave mine up for context, but I recommend using Lee's suggestion.

Author: Good point, thanks

> * Creates the second generation backing index, name `.ds-timeseries-000002`.

Reviewer: Typo
```suggestion
* Creates the second generation backing index, named `.ds-timeseries-000002`.
```

> As it is the backing index of the `timeseries` data stream which was created using the `timeseries_template`
> the template configuration will be applied to the new index.

Reviewer: Nit: Reword for clarity
```suggestion
Because it is a backing index of the `timeseries` data stream, the configuration from the `timeseries_template` composable template is applied to the new index.
```

* As it is the latest generation index of the `timeseries` data stream, the newly created
backing index `.ds-timeseries-000002` becomes the data stream's write index.

> This process repeats each time rollover conditions are met.

Reviewer: Fix for OR logic
```suggestion
This process repeats each time a rollover condition is met.
```

You can search across all of the data stream's backing indices, managed by the `timeseries_policy`,
with the `timeseries` data stream name.

> Write operations are routed to the current write index, whilst read operations will be handled by all
> backing indices.

Reviewer: Breaks this into two sentences for readability.
```suggestion
Write operations are routed to the current write index. Read operations will be handled by all
backing indices.
```

[discrete]
[[ilm-gs-check-progress]]
=== Check lifecycle progress

To get status information for managed indices, you use the {ilm-init} explain API.
This lets you find out things like:

* What phase an index is in and when it entered that phase.
* The current action and what step is being performed.
* If any errors have occurred or progress is blocked.

> For example, the following request gets information about the `timeseries` indices:

Reviewer: ```suggestion
For example, the following request gets information about the `timeseries` data stream's
backing indices:
```

[source,console]
--------------------------------------------------
GET .ds-timeseries-*/_ilm/explain
--------------------------------------------------
// TEST[continued]

> The response below shows that the data stream's first generation backing index is waiting in the `hot`

Reviewer: ```suggestion
The following response shows the data stream's first generation backing index is waiting for the `hot`
```

phase's `rollover` action.

> It remains in this state and {ilm-init} continues to call `attempt-rollover` until the rollover conditions
> are met.

Reviewer: OR logic.
```suggestion
It remains in this state and {ilm-init} continues to call `attempt-rollover` until a rollover condition
is met.
```

// [[36818c6d9f434d387819c30bd9addb14]]
[source,console-result]
--------------------------------------------------
{
  "indices": {
    ".ds-timeseries-000001": {
      "index": ".ds-timeseries-000001",
      "managed": true,
      "policy": "timeseries_policy",             <1>
      "lifecycle_date_millis": 1538475653281,
      "age": "30s",                              <2>
      "phase": "hot",
      "phase_time_millis": 1538475653317,
      "action": "rollover",
      "action_time_millis": 1538475653317,

>       "step": "attempt-rollover",                <3>

Reviewer 2: Super minor, but more likely to be:
```suggestion
      "step": "check-rollover-ready",                <3>
```

      "phase_execution": {
        "policy": "timeseries_policy",
        "phase_definition": {                    <4>
          "min_age": "0ms",
          "actions": {
            "rollover": {
              "max_size": "50gb",
              "max_age": "30d"
            }
          }
        },
        "version": 1,
        "modified_date_in_millis": 1539609701576
      }
    }
  }
}
--------------------------------------------------
// TESTRESPONSE[skip:no way to know if we will get this response immediately]

<1> The policy used to manage the index
<2> The age of the index
<3> The step {ilm-init} is performing on the index
<4> The definition of the current phase (the `hot` phase)

//////////////////////////

[source,console]
--------------------------------------------------
DELETE /_data_stream/timeseries
--------------------------------------------------
// TEST[continued]

//////////////////////////

//////////////////////////

[source,console]
--------------------------------------------------
DELETE /_index_template/timeseries_template
--------------------------------------------------
// TEST[continued]

//////////////////////////

[discrete]

> [[ilm-gs-create-policy]]

Reviewer: This is breaking the docs CI test. I'll go ahead and merge this change in.
```suggestion
[[manage-time-series-data-without-data-streams]]
```

Author: Thanks for fixing this James. It wasn't obvious what the offending piece was from the test failure log output

=== Manage time-series data without data streams

Even though <<data-streams, data streams>> are a convenient way to scale
and manage time-series data, they are designed to be append-only. We recognise there
might be use-cases where data needs to be updated or deleted in place and the

> <<data-streams, data streams>> APIs don't support the delete and update APIs directly,

Reviewer: This could be misread to state that you can't delete a data stream. Reword for clarity.
```suggestion
data streams don't support the delete and update requests directly,
```

so the index APIs would need to be used directly on the data stream's backing indices.

> Another option to manage the time-series data in these cases is to use an index alias
> so you can periodically roll over to a new index.

Reviewer: Nit: Reword for clarity.
```suggestion
In these cases, you can use an index alias to manage indices containing the time-series data
and periodically roll over to a new index.
```

> To automate rollover and management of time-series indices with {ilm-init} over an index

Reviewer: ```suggestion
To automate rollover and management of time-series indices with {ilm-init} using an index
```
